zerver/tests/test_subs.py

```diff
@@ -1625,6 +1625,76 @@ def test_create_stream_policy_setting(self) -> None:
+        """
+        user_profile = self.example_user('hamlet')
+        user_profile.date_joined = timezone_now()
+        user_profile.save()
+        self.login_user(user_profile)
+        do_change_user_role(user_profile, UserProfile.ROLE_MEMBER)
+
+        # Allow all members to create streams.
+        do_set_realm_property(user_profile.realm, 'create_private_stream_policy',
+                              Realm.POLICY_MEMBERS_ONLY)
+        # Set waiting period to 10 days.
+        do_set_realm_property(user_profile.realm, 'waiting_period_threshold', 10)
+
+        # Can successfully create stream despite being less than waiting period and not an admin,
+        # due to create stream policy.
+        stream_name = ['all_members']
+        result = self.common_subscribe_to_streams(user_profile, stream_name, invite_only=True)
+        self.assert_json_success(result)
+
+        # Allow only administrators to create streams.
+        do_set_realm_property(user_profile.realm, 'create_private_stream_policy',
+                              Realm.POLICY_ADMINS_ONLY)
+
+        # Cannot create stream because not an admin.
+        stream_name = ['admins_only']
+        result = self.common_subscribe_to_streams(user_profile, stream_name, invite_only=True, allow_fail=True)
+        self.assert_json_error(result, 'User cannot create private streams.')
+
+        # Make current user an admin.
+        do_change_user_role(user_profile, UserProfile.ROLE_REALM_ADMINISTRATOR)
+
+        # Can successfully create stream as user is now an admin.
+        stream_name = ['admins_only']
+        self.common_subscribe_to_streams(user_profile, stream_name, invite_only=True)
+
+        # Allow users older than the waiting period to create streams.
+        do_set_realm_property(user_profile.realm, 'create_private_stream_policy',
+                              Realm.POLICY_FULL_MEMBERS_ONLY)
+
+        # Can successfully create stream despite being under waiting period because user is admin.
+        stream_name = ['waiting_period_as_admin']
+        self.common_subscribe_to_streams(user_profile, stream_name, invite_only=True)
+
+        # Make current user no longer an admin.
+        do_change_user_role(user_profile, UserProfile.ROLE_MEMBER)
+
+        # Cannot create stream because user is not an admin and is not older than the waiting
+        # period.
+        stream_name = ['waiting_period']
+        result = self.common_subscribe_to_streams(user_profile, stream_name, invite_only=True, allow_fail=True)
+        self.assert_json_error(result, 'User cannot create private streams.')
+
+        # Make user account 11 days old..
+        user_profile.date_joined = timezone_now() - timedelta(days=11)
+        user_profile.save()
+
+        # Can successfully create stream now that account is old enough.
+        stream_name = ['waiting_period']
+        self.common_subscribe_to_streams(user_profile, stream_name, invite_only=True)
```

Comment:
It would be nice if you can figure out a way to make this nice, large test not duplicated using a loop or calling the same function twice, e.g. having a `subscribe_kwargs` property containing `invite_only` or not.

Reply:
(Ideally, this would be done as a prep commit that we can merge without reading the rest, that just refactors this test function, and then the second commit would just call it twice with different arguments, so that we can merge that refactoring and forget about it while resolving whatever other issues are in this PR -- one of Zulip's strategies for making review more efficient)

Reply:
But then it won't be individually deployable, right? Since it tests properties that don't exist yet.

Reply:
Well, you can do a prep commit that just refactors the existing code to have a structure that can support the parameter, and have the existing test call it that way -- and then the second commit just adds the new test function that calls the shared function with different `subscribe_kwargs` or whatever.